### PR TITLE
#4147 added `tenant_id` tag to tracing spans

### LINKED ISSUE #4147: Add tracing to the tenant federation mergeQueryable
This is a work item that came out the review of https://github.com/cortexproject/cortex/pull/4117

The mergeQueryable aggregates data from underlying queryables concurrently. Understanding the call sequence is only possible providing spans.

/assign @simonswine 

Some examples that creates spans:

https://github.com/cortexproject/cortex/blob/d382e1d80eaf0847e3b57592b970a4f1bf3121df/pkg/querier/blocks_store_queryable.go#L318

https://github.com/cortexproject/cortex/blob/5009276fdd385587a13f1492520cc38647512bc6/pkg/querier/worker/scheduler_processor.go#L134

### PR DESCRIPTION
**Pull request by vlad-diachenko**
**What this PR does**:
1. added logic to add tag `tenant_id` to tracing spans if tenant is defined in context.
2. added additional tracing spans to merge_querier's methods

**Which issue(s) this PR fixes**:
Fixes #4147

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/spanlogger/spanlogger.go
Comment (simonswine) [inline pkg/util/spanlogger/spanlogger.go:43]: I think it makes sense to use the slice directly instead of a string value and then call the tag "tenant_ids"

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (simonswine) [inline pkg/querier/tenantfederation/merge_queryable_test.go:23]: I think I wouldn't rename the import necessarily. What is your thinking there?

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (simonswine) [inline pkg/querier/tenantfederation/merge_queryable.go:102]: I think it would make sense to create a span during the creation of the queriers and the finish the span once `Close()` of the resulting mergeQuerier is called.

I have played with that here: 

https://github.com/cortexproject/cortex/commit/6ebd273bf665752794843aa02495b8a6ede0f15f

That's how that looks like:
scrn-2021-05-17-12-46-02

File: pkg/util/spanlogger/spanlogger.go
Comment (simonswine) [inline pkg/util/spanlogger/spanlogger.go:42]: I am not too sure what other side effects this could have, but I think in general it is a good idea to always have the context's tenant_id(s) in the tracing span.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (simonswine) [reply pkg/querier/tenantfederation/merge_queryable.go:102]: This feels at least like it can group the same tenants nicely together. Obvisouly for a real fix we need context propagation as a part of LabelValues,LabelName, Select.

**Review by simonswine — COMMENTED** [review]
Thanks this looks already quite good. :+1: 

See my comments in-line and feel free to ping me when things are unclear

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:102]: yes, it will help to group spans by tenant_id . thanks, i will implement it

File: pkg/querier/tenantfederation/merge_queryable_test.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable_test.go:23]: done

File: pkg/util/spanlogger/spanlogger.go
Comment (vlad-diachenko) [reply pkg/util/spanlogger/spanlogger.go:43]: make sense . done

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:102]: however, it might confuse a little bit because all spans `mergeQuerier.NewQuerier` will take almost similar time for each tenant. but inside this span we will see real picture. 
in current implementation it looks less confusing, i think. 
image

what do you think?

Comment (simonswine) [reply pkg/querier/tenantfederation/merge_queryable.go:102]: To be honest I agree partly with your point, but I am not too sure, how to fix this properly. 

I guess part of the problem is that the a query is only finished once `Close()` is called, which happens at the same time for underlying queriers.

I think the most confusing part of not grouping spans by tenant_id is seeing all (and even unrelated) spans together. I think until we implement context passing in those methods, we won't be able to solve that properly.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (replay) [inline pkg/querier/tenantfederation/merge_queryable.go:140]: to avoid confusion I'd keep the case of the method name (`L` instead of `l`)
```suggestion
	log, _ := spanlogger.New(m.ctx, "mergeQuerier.LabelValues")
```

Same for the other two below

File: pkg/util/spanlogger/spanlogger.go
Comment (replay) [inline pkg/util/spanlogger/spanlogger.go:42]: instead of `ids != nil` i'd check `len(ids) > 0`, because if `ids == nil` then `len(ids) == 0` but `ids` could also be `[]string{}` which is `!= nil`.
```suggestion
	if ids, _ := tenant.TenantIDs(ctx); len(ids) > 0 {
```

https://play.golang.org/p/zO8541dX4xN

File: pkg/util/spanlogger/spanlogger.go
Comment (replay) [inline pkg/util/spanlogger/spanlogger.go:19]: should this maybe be plural? since you're setting a `[]string` value
```suggestion
	TenantIDTagName = "tenant_ids"
```

**Review by replay — APPROVED** [review]
LGTM, I only have minor comments

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:140]: done

File: pkg/util/spanlogger/spanlogger.go
Comment (vlad-diachenko) [reply pkg/util/spanlogger/spanlogger.go:42]: done

File: pkg/util/spanlogger/spanlogger.go
Comment (vlad-diachenko) [reply pkg/util/spanlogger/spanlogger.go:19]: done

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (gouthamve) [reply pkg/querier/tenantfederation/merge_queryable.go:102]: You could instantiate the queriers (call the callback inside LabelValues, etc.) and the right context will be applied.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (gouthamve) [inline pkg/querier/tenantfederation/merge_queryable.go:163]: This is interesting because the `mergeDistinctStringSlice` uses `m.ctx` internally. We should pass the context to this function to make it use the right one.

Comment (gouthamve) [reply pkg/querier/tenantfederation/merge_queryable.go:163]: But just noticed that it wouldn't make much difference anyways.

File: pkg/util/spanlogger/spanlogger.go
Comment (gouthamve) [inline pkg/util/spanlogger/spanlogger.go:42]: This is great!

**Review by gouthamve — APPROVED** [review]
LGTM, and I wonder how bad is creating the `queriers` inside `Select`, `LabelValues`, etc.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:163]: yes, it won't make much difference because the main pain point that queriers use context passed during creation. but anyway, nice catch. i fix it now to not fix it in future ;)

**Review by simonswine — APPROVED** [review]
LGTM

I think there is still a `go mod vendor` necessary to make the CI/CD work

**Comment by simonswine** [comment]
> I wonder how bad is creating the `queriers` inside `Select`, `LabelValues`, etc.

I think with `Select` it might be problematic, as it would have to `Close()` before the `stoarge.SeriesSet` is fully iterated through

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:23]: Also please move it above, along with the group of `[ENHANCEMENT]` entries.

```suggestion
* [ENHANCEMENT] Added `tenant_ids` tag to tracing spans. #4186
```

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pracucci) [inline pkg/querier/tenantfederation/merge_queryable.go:153]: [nit] Please rollback this change.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pracucci) [inline pkg/querier/tenantfederation/merge_queryable.go:164]: [nit] Please rollback this change.

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pracucci) [inline pkg/querier/tenantfederation/merge_queryable.go:209]: [nit] Single line, please.

**Review by pracucci — APPROVED** [review]
LGTM modulo a couple of nits I would be glad to see fixed before merging. Thanks!

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:209]: done

File: CHANGELOG.md
Comment (vlad-diachenko) [reply CHANGELOG.md:23]: done

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:164]: you are right, this context is not used in callback function so it's not necessary to pass it to this method. done

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go:153]: you are right, this context is not used in callback function so it's not necessary to pass it to this method. done

File: pkg/querier/tenantfederation/merge_queryable.go
Comment (pracucci) [inline pkg/querier/tenantfederation/merge_queryable.go:152]: Sorry for being unclear! It's perfectly fine to pass `ctx` to `mergeDistinctStringSlice()`. I just meant to rollback the splitting to 2 lines.

Comment (vlad-diachenko) [reply pkg/querier/tenantfederation/merge_queryable.go]: done

File: pkg/util/spanlogger/spanlogger.go
Comment (pstibrany) [inline pkg/util/spanlogger/spanlogger.go:42]: This is going to add tenant tag to every span created via this method. That seems unnecessary.

Comment (pracucci) [reply pkg/util/spanlogger/spanlogger.go:42]: It's true that's unnecessary, but is that a bad thing?

Comment (pracucci) [reply pkg/util/spanlogger/spanlogger.go:42]: Alternatively, we can add functional options to a `NewWithLoggerWithOptions()` function an conditionally enable it only when the option is specified. What do you think?

Comment (pstibrany) [reply pkg/util/spanlogger/spanlogger.go:42]: > Alternatively, we can add functional options to a `NewWithLoggerWithOptions()` function an conditionally enable it only when the option is specified. What do you think?

You've convinced me that it's ok. I would keep it as is.

Comment (vlad-diachenko) [reply pkg/util/spanlogger/spanlogger.go:42]: > Alternatively, we can add functional options to a NewWithLoggerWithOptions() function an conditionally enable it only when the option is specified. What do you think?

it was the first implementation when I added separate method to create a span with tenant tag. however, all existing span creations had to be revised and some of them had to be replaced with a new method. otherwise tenant_ids won't be defined for some spans when we investigate something on tracing dashboard. 
I found that current implementation the easiest way and much more safe because when somebody add a new code he won't forget to create a span with tenant tag.

> This is going to add tenant tag to every span created via this method. That seems unnecessary.
unfortunately, yes. if it's critical we can think how to add this tag only if `tenant_federation.enabled` is true.

Comment (pstibrany) [reply pkg/util/spanlogger/spanlogger.go:42]: > @pstibrany
> unfortunately, yes. if it's critical we can think how to add this tag only if `tenant_federation.enabled` is true.

After some thinking and looking at our existing spans, I think it's fine... one extra tag per span is no big deal, and one can immediately see tenant when checking the trace.

Comment (vlad-diachenko) [reply pkg/util/spanlogger/spanlogger.go:42]: @pstibrany fully agree with you ;) thanks for your review

Comment (pracucci) [reply pkg/util/spanlogger/spanlogger.go:42]: It's always good to have conversations! 💪

**Pull request merged**
